Be type-strict checking on equality checks

hpvb opened the pull request:
After a short discussion with @reduz and @karroffel we decided to make
all non number/number comparisons return type errors on comparisons.

Now bool == bool is allowed but Vector2 == Vector3 is a type error and
no longer 'not equal'. The same has been done for the != operators.

In addition I forgot to add some failures to some Object operators
meaning that there was a potential for a crasher.

This PR also includes a fix for booleanization pointed out by @jjay. The code
```gdscript
var x = Dictionary()
if not x:
  print("empty dict")
```
is now valid GDScript

This fixes #11411

hpvb committed:
Be type-strict checking on equality checks

After a short discussion with @reduz and @karroffel we decided to make
all non number/number comparisons return type errors on comparisons.

Now bool == bool is allowed but Vector2 == Vector3 is a type error and
no longer 'not equal'. The same has been done for the != operators.

In addition I forgot to add some failures to some Object operators
meaning that there was a potential for a crasher.

jjay commented:
GDScript is dynamically typed programming language. With this PR it will be less dynamically typed. 

Now I feel uncomfortable even because of I can't use arrays/dict in boolean context (`if some_dict: pass`), but this PR will force me to make double check for equality (`if typeof(a) == typeof(b) && a == b: pass`). It wold be nice to put the IRC discussion here. I completely don't understand the advantage of failing over returning false. Python btw is ok with comparing different types.

akien-mga commented:
AFAIU you can still do `if some_dict: pass`, but no longer `if some_dict == true: pass`.

jjay commented:
In 2.1 I get `cannot evaluate conditional expression of type: Dictionary` for `if dict: pass`

A little bit more hate to this pr (first line from gdscript tutorial):
```gdscript
var dict = {"key":"value", 2:3}
for key in dict:
  if key == "key":
    print("Please let me do eq checks wherever I want!")
```

jjay commented:
Ok, I double check current behaviour (for 2.1) and its not let me compare different types for equality - `Invalid operands 'int' and 'String' in operator '=='`. I'm surprised how I missed this, but this is completely discourages and I don't understand why i cant compare =/

karroffel commented:
@jjay it doesn't become any less dynamically typed, it only becomes more strongly typed.

For your example you can use `match`.

```gdscript
var dict = {"key":"value", 2:3}

for k in dict:
    match k:
        "key": print("hey cool, this works!")
        2:     print("this works too!")
        _:
            match typeof(k):
                TYPE_FLOAT: print("it's a float")
                TYPE_DICT:  print("it's another dictionary")
                _:          print("It's something else.")
```

`match` handles those type checks for you.

hpvb commented:
> GDScript is dynamically typed programming language. With this PR it will meke it less dynamically typed.

Yes, but GDScript's variables are dynamic, but it is strongly typed. There's a difference between allowing variables to dynamically change type and having a weak typing system. 

> In 2.1 I get cannot evaluate conditional expression of type: Dictionary for if dict: pass
I think that Dictionary has a working Booleanize() if it doesn't I'll add one. I think it makes sense to be able to tell if dict is empty.

>  if k == "key":

(see @karroffel 's example)

> Ok, I double check current behaviour (for 2.1) and its not let me compare different types for equality

The main problem was that in 2.1 this behavior is quite random. some things can be checked for equality if one type is on the left hand side of the comparison, some can't. it's a mess.

hpvb committed:
Allow booleanization of all types

We now allow booleanization of all types. This means that empty versions
of all types now evaluate to false. So a Vector2(0,0), Dictionary(),
etc.

This allows you to write GDScript like:
if not Dictionary():
  print("Empty dict")

Booleanization can now also no longer fail. There is no more valid flag,
this changes Variant and GDNative API.

akien-mga commented on core/variant_op.cpp:
```diff
@@ -161,35 +162,78 @@ bool Variant::booleanize(bool &r_valid) const {
-		case COLOR:
+		case VECTOR2: {
+			const Vector2 *v = reinterpret_cast<const Vector2 *>(_data._mem);
+			return v->x + v->y;
```
As discussed on IRC, this would make `Vector2(1, -1)` or `Vector2(-42, 42)` evaluate to `true`. The logic should be: false if all components are 0, else true, so `v->x || v->y`.

akien-mga commented on core/variant_op.cpp:
```diff
@@ -161,35 +162,78 @@ bool Variant::booleanize(bool &r_valid) const {
+		}
+		case VECTOR3: {
+			const Vector3 *v = reinterpret_cast<const Vector3 *>(_data._mem);
+			return v->x + v->y + v->z;
```
Same here, `v->x || v->y || v->z`.

akien-mga commented on core/variant_op.cpp:
```diff
@@ -161,35 +162,78 @@ bool Variant::booleanize(bool &r_valid) const {
+		}
+		case RECT2: {
+			const Rect2 *v = reinterpret_cast<const Rect2 *>(_data._mem);
+			return v->position.x + v->position.y + v->size.x + v->size.y;
```
That's also wrong logic, but what should be done instead is up for discussion. What would be a "false" Rect2?
I guess we have two possibility:
- `position` is false (Point2(0,0)) and `size` is false (Vector2(0,0))
- `size` is false, regardless of the position

I guess the former is likely more consistent (like for other types, there would be only one false case), even if one could argue that an empty Rect2 at position (100, 213) could be "false".

akien-mga commented on core/variant_op.cpp:
```diff
@@ -161,35 +162,78 @@ bool Variant::booleanize(bool &r_valid) const {
+		}
+		case TRANSFORM2D: {
+			const Transform2D *v = _data._transform2d;
+			return v->elements[0].x + v->elements[0].y +
```
Same here, should be false if all components are null. I guess you could do `booleanize(elements[0]) || booleanize(elements[1]) || booleanize(elements[2])`, though maybe that would be less efficient than just hardcoding the six checks :)

akien-mga commented on core/variant_op.cpp:
```diff
@@ -161,35 +162,78 @@ bool Variant::booleanize(bool &r_valid) const {
 		case DICTIONARY:
+			return !(*reinterpret_cast<const Dictionary *>(_data._mem)).empty();
 		case ARRAY:
+			return !(*reinterpret_cast<const Array *>(_data._mem)).empty();
```
For the other *Array types you used `*Array.size()`, is there a difference in cost using `!Array.empty()` here?

akien-mga commented on core/variant_op.cpp:
```diff
@@ -161,35 +162,78 @@ bool Variant::booleanize(bool &r_valid) const {
 		case NODE_PATH:
 			return (*reinterpret_cast<const NodePath *>(_data._mem)) != NodePath();
 		case DICTIONARY:
+			return !(*reinterpret_cast<const Dictionary *>(_data._mem)).empty();
```
Looks like `Dictionary.size()` would be a tad better: https://github.com/godotengine/godot/blob/master/core/dictionary.cpp#L122-L129

akien-mga commented on core/variant_op.cpp:
```diff
@@ -161,35 +162,78 @@ bool Variant::booleanize(bool &r_valid) const {
 		case OBJECT:
 			return _get_obj().obj;
 		case NODE_PATH:
 			return (*reinterpret_cast<const NodePath *>(_data._mem)) != NodePath();
```
Nitpick, but maybe `!NodePath.is_empty()` would be better?
https://github.com/godotengine/godot/blob/master/core/node_path.cpp#L406-L409

LikeLakers2 commented:
Personally I don't get this. What's the benefit to this? How does this benefit users in any good way? What point does this PR serve? And as was pointed out before: What's the point of erroring rather than returning false?

I'm not gonna argue about whether this makes it more or less strongly or dynamically typed. I just wanna know what the reasoning is that two objects of different types cannot just return false when you `==` them.

hpvb commented:
I'm beginning to feel that I named this PR very poorly. In the previous situation comparing two values of different types was a type error *most of the time*. After this PR it is *always* a type error.

Previously you could do 
```gdscript
x = Dictionary()
if x == 1:
  print("wut?")
```
But NOT
```gdscript
x = Dictionary()
if 1 == x:
  <typerror>
```

This really only makes things more consistent, it really shouldn't break any existing code.

hpvb commented:
As for what it's FOR, by tidying up the type system we open up possibilities for much more advanced optimization techniques. We want to be able to compile GDScript to a fully typed VM in order to make it much faster than it is now. 

This change does not seemingly directly benefit users but in the long run (3.1, 3.2 maybe) every GDScript user should see a huge increase in GDScript performance.

LikeLakers2 commented:
Again then, I'll ask. Why don't we just return false if the two objects given are different types? If they're different types, they're already different as far as the program is concerned, so even if it's something dumb like `if null == false` then it should still return false.

hpvb commented:
Once GDScript gets compiled there won't be runtime typechecking anymore. That's kind of the point.

LikeLakers2 commented:
So, just so I'm getting this straight, the answer to my question is not "we don't want to" (as I thought it was), but moreso "we can't"?

karroffel commented:
We totally could if we wanted, but it's "we have plans for the future so we don't want to introduce those things now as they will make everything else harder".
<br/>
<br/>
<br/>
<br/>
<br/>
<br/>
<br/>
<br/>
<br/>
<br/>
Also comparing values of different incompatible types makes no sense at all, so returning false is worse than an error since it let's you write sluggish code and get away with it.

hpvb commented:
It would make future optimizations harder to do, and it would lead to slow and fast ways of doing things once we have those optimizations. Because we don't want to let people write code now that will be slow in the future we're just not creating a new feature that can really only be used to write slow code now.

Again, GDScript *never* let you do things like if "1" == Vector2(1,1): that was always a type error.

All this PR does is make that consistent across all types. 

If your questions is "Will you add type checking between different types for 3.0 as a new feature" the answer is "no, because it would cost us some optimization opportunities in the future that we'd like to keep open."

Does that answer your questions? :)

LikeLakers2 commented:
@karroffel Not sure I really agree that it's *worse*. Yeah you're right, it does let you write sluggish code, but at that point you really only have the person writing said code to blame, not Godot or Godot's devs. I admit I've done it myself before, but the difference is that it always came back to bite me in the butt later. I understand the point, though, and if I am to assume that you understand my question as I do, then I also understand the reasoning for not doing this.

@hpvb The question, again, is "Why do comparisons between different types not return false?" Or, to word it as a yes/no question, "Will you change comparisons to return false instead of TypeError when comparing different types?" Does that help a little bit?

-----

To both of you: I totally get if it's something you want to leave out for now, but it's something I feel should definitely be changed in the future. Whether it allows for sloppy code or not, it prevents us from having to write both `if typeof(j) == typeof(k)` *and* `if j == k`, instead condensing it down into `if j == k`.

hpvb commented:
@LikeLakers2 The answer to will you is "No" The answer to why is "It will make writing a faster GDScript runtime later harder and it may never be as fast as if we don't allow these types of operations at all."

akien-mga commented:
> To both of you: I totally get if it's something you want to leave out for now, but it's something I feel should definitely be changed in the future. Whether it allows for sloppy code or not, it prevents us from having to write both if typeof(j) == typeof(k) and if j == k, instead condensing it down into if j == k.

But that was never allowed in the first place.

In Godot 2.1.4:

This compiles fine: 
![image](https://user-images.githubusercontent.com/4701338/30608982-5b108478-9d7a-11e7-8f3f-53898c29f1f1.png)

Run your game, and hey, here comes the friendly frustration:
![image](https://user-images.githubusercontent.com/4701338/30609005-68a7b016-9d7a-11e7-8553-deb5ebd8b90b.png)

This does not compile:
![image](https://user-images.githubusercontent.com/4701338/30609110-c0c48454-9d7a-11e7-8d05-76274d0aa217.png)

It's exactly the same in the master branch after this PR. But thanks to this PR, we can catch the error at compilation time in the future.

vnen commented:
The thing is that rarely (if ever) you'd need to do `if typeof(j) == typeof(k)` because most of the time you do comparisons with the same type anyway. Throwing an error when different types are compared would indicate a flaw in your logic and would be an opportunity to fix your code, while returning `false` would just confuse you.

Also, returning `false` always when the types are different is not really user-friendly, because type coercion would be expected in such case. For instance, type coercion would make `true` expressions like `null == 0` and `false == ""`, since they are all "falsy" values (as they say in JavaScript).

I prefer strict types, they leave less margin for errors.

karroffel commented:
> Whether it allows for sloppy code or not, it prevents us from having to write both `if typeof(j) == typeof(k)` and `if j == k`, instead condensing it down into `if j == k`.

We hope to add

 1) optional type hints and
 2) global type inference

in the future, so wouldn't need those runtime type checks, the compiler would tell you if there are problems in your code before running it.

-----

> Yeah you're right, it does let you write sluggish code, but at that point you really only have the person writing said code to blame, not Godot or Godot's devs.

Okay, you can use GDNative to write your code in assembler, I mean it gives you *all* the flexibility in the world and you're the only one to blame, so why even bother with language design? 😛 😉 

I think the other people here have said it pretty nicely already, I'm just going to agree to them. 👍
